File: AndroidNativeAudio/app/src/main/java/com/example/simulo/androidnativeaudio/Music.java
package com.example.simulo.androidnativeaudio;

import android.app.Activity;

import java.util.ArrayList;
import java.util.List;

public class Music {
    static Activity ParentActivity;
    static List<MusicPlayer> players = new ArrayList<>();

    static void init(Activity activity) {
        ParentActivity = activity;
    }

    public static int loadAsync(String assetPath)
    {
        int musicId;

        try {
            MusicPlayer player = new MusicPlayer(assetPath, ParentActivity);
            player.loadAsync();
            players.add(player);
            musicId = players.indexOf(player);
        } catch (Exception e) {
            return -1;
        }

        return musicId;
    }

    public static boolean isLoadCompleted(int musicId) {
        return players.get(musicId).isLoadCompleted();
    }

    public static int play(int musicId) {
        try {
            players.get(musicId).play();
        } catch (Exception e) {
            return -1;
        }

        return 0;
    }

    public static int stop(int musicId) {
        try {
            players.get(musicId).stop();
        } catch (Exception e) {
            return -1;
        }

        return 0;
    }

    public static int setLooping(int musicId, boolean isLooping) {
        try {
            players.get(musicId).setLooping(isLooping);
        } catch (Exception e) {
            return -1;
        }

        return 0;
    }

    public static boolean isLooping(int musicId) {
        boolean result;
        try {
            result = players.get(musicId).isLooping();
        } catch (Exception e) {
            return false;
        }

        return result;
    }

    public static int setVolume(int musicId, float volume) {
        try {
            players.get(musicId).setVolume(volume);
        } catch (Exception e) {
            return -1;
        }

        return 0;
    }

    public static int release(int musicId) {
        try {
            players.get(musicId).release();
            players.remove(musicId);
        } catch (Exception e) {
            return -1;
        }

        return 0;
    }

    public static void release() {
        for (MusicPlayer player : players)
            player.release();
    }
}

File: AndroidJniAudio/src/Music.cpp
#include "Music.h"
#include "Audio.h"
#include "Logger.h"
#include <SDL2/SDL.h>

namespace sb
{
	void Music::load(std::string assetPath)
	{
		#ifdef __ANDROID__
			if (!Audio::isInit())
				Audio::init();

			m_assetPath = assetPath;
			validateFileEnding(assetPath);

			JNI_WARN(m_id = sb::Java::callStaticIntMethod("org/libsdl/app/Music", "loadAsync", "(Ljava/lang/String;)I", sb::Java::newUtfString(assetPath)))
			
			jint loadResult = 0;
			while (loadResult == jint(0)) {
				JNI_WARN(loadResult = sb::Java::callStaticIntMethod("org/libsdl/app/Music", "isLoadComplete", "(I)I", m_id));
				SDL_Delay(1);
			}
		#endif
	}

	void Music::play()
	{
		#ifdef __ANDROID__
			JNI_WARN(sb::Java::callStaticIntMethod("org/libsdl/app/Music", "play", "(I)I", m_id));
		#endif
	}

	void Music::stop()
	{
		#ifdef __ANDROID__
			JNI_WARN(sb::Java::callStaticIntMethod("org/libsdl/app/Music", "stop", "(I)I", m_id));
		#endif
	}

	void Music::setLooping(bool looping)
	{
		#ifdef __ANDROID__
			JNI_WARN(sb::Java::callStaticIntMethod("org/libsdl/app/Music", "setLooping", "(IZ)I", m_id, jboolean(looping)));
		#endif
	}

	void Music::setVolume(float volume)
	{
		#ifdef __ANDROID__
			JNI_WARN(sb::Java::callStaticIntMethod("org/libsdl/app/Music", "setVolume", "(IF)I", m_id, jfloat(volume)));
		#endif
	}

	void Music::validateFileEnding(std::string assetPath)
	{
		std::size_t pos = assetPath.rfind(".");    
		std::string ending = assetPath.substr(pos, std::string::npos);

		sb::Logger().errorIf(ending != ".ogg") 
			<< "Music files must be in .ogg format" << std::endl;
	}
}
